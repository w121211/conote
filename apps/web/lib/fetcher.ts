import { resolve } from 'path'
import { FetchClient } from './fetcher/fetcher'

const fetchClientPropertyName = `__prevent-name-collision__fetcher`

type GlobalThisWithFetchClient = typeof globalThis & {
  [fetchClientPropertyName]: FetchClient
}

function getFetchClient(): FetchClient {
  if (process.env.NODE_ENV === `production`) {
    return new FetchClient()
  } else {
    const newGlobalThis = globalThis as GlobalThisWithFetchClient
    if (!newGlobalThis[fetchClientPropertyName]) {
      newGlobalThis[fetchClientPropertyName] = new FetchClient(resolve(process.cwd(), '.fetcher_local_cache.dump.json'))
    }
    return newGlobalThis[fetchClientPropertyName]
  }
}

const fetcher = getFetchClient()

export default fetcher
