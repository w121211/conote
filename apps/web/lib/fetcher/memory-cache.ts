/**
 * A simple memory cache, mainly use for avoiding repeatly calls on the same urls during development
 */
import { readFileSync, writeFileSync } from 'fs'
import { LinkScrapedResult } from '../interfaces'

export class MemoryCache {
  private filepath: string
  private cache: Record<string, LinkScrapedResult> = {}

  constructor(cachePath: string) {
    this.filepath = cachePath
    try {
      this.cache = this.load(this.filepath)
    } catch (err) {
      console.log(`cache not found, create a new one`)
      //   console.log('No ')
    }
  }

  private load(filepath: string): Record<string, LinkScrapedResult> {
    console.log(`loading cache from ${filepath}`)
    return JSON.parse(readFileSync(filepath, { encoding: 'utf8' }))
  }

  public dump(): void {
    console.log(`Dump cache to ${this.filepath}`)
    writeFileSync(this.filepath, JSON.stringify(this.cache))
  }

  public get(url: string): LinkScrapedResult {
    if (url in this.cache) {
      console.log('hit cache: ' + url)
      return this.cache[url]
    }
    throw new Error('Cache not found:' + url)
  }

  public set(url: string, result: LinkScrapedResult): void {
    this.cache[url] = result
  }
}
