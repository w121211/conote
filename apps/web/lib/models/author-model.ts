import { Author } from '@prisma/client'
import {
  Author as GQLAuthor,
  AuthorMeta as GQLAuthorMeta,
  AuthorInput as GQLAuthorInput,
} from 'graphql-let/__generated__/__types__'
import prisma from '../prisma'

// type Job = {
//   name: string
//   org: string
//   startedAt?: Date
//   endedAt?: Date
// }

type AuthorMeta = {
  type: 'ORG' | 'PERSON'
  job?: string // eg youtuber, analylist
  org?: string
  sites: [string, string][] // [site_url, site_name]
}

export const AuthorModel = {
  async get(id?: string, name?: string): Promise<GQLAuthor | null> {
    let author: Author | null
    if (id) {
      author = await prisma.author.findUnique({
        where: { id },
      })
    } else if (name) {
      author = await prisma.author.findUnique({
        where: { name },
      })
    } else {
      throw 'Need either id or name to get author'
    }
    if (author) {
      return this.toGQLAuthor(author)
    }
    return null
  },

  async getAll(): Promise<Author[]> {
    console.log('Retreiving all authors from database...')

    let authors: Author[] = []
    let cursor: string | undefined = undefined

    // eslint-disable-next-line no-constant-condition
    while (true) {
      // eslint-disable-next-line no-await-in-loop
      const res: Author[] = await prisma.author.findMany({
        take: 100,
        skip: cursor ? 1 : undefined, // skip cursor itself
        orderBy: { id: 'asc' },
        cursor: cursor ? { id: cursor } : undefined,
      })
      if (res.length === 0) {
        break
      }
      authors = authors.concat(res)
      cursor = res[res.length - 1].id
    }
    return authors
  },

  async create(input: GQLAuthorInput): Promise<Author> {
    const { name, type, job, org, sites } = input
    const meta: AuthorMeta = {
      type,
      job: job ?? undefined,
      org: org ?? undefined,
      sites: sites.map(e => [e.url, e.name]),
    }
    return prisma.author.create({
      data: {
        name,
        meta,
      },
    })
  },

  toGQLAuthor(author: Author): GQLAuthor {
    const { meta } = author
    return {
      ...author,
      meta: this.toGQLAuthorMeta(meta as unknown as AuthorMeta),
    }
  },

  toGQLAuthorMeta(meta: AuthorMeta): GQLAuthorMeta {
    const { sites } = meta
    return {
      ...meta,
      sites: sites.map(([url, name]) => {
        return { url, name }
      }),
    }
  },

  async update(id: string, input: GQLAuthorInput): Promise<Author> {
    const { name, type, job, org, sites } = input
    const meta: AuthorMeta = {
      type,
      job: job ?? undefined,
      org: org ?? undefined,
      sites: sites.map(e => [e.url, e.name]),
    }
    return prisma.author.update({
      data: {
        name,
        meta,
      },
      where: { id },
    })
  },
}
