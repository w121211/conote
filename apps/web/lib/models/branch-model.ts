if (process.env.APP_BRANCH_NAME === undefined) {
  throw new Error('Not found APP_BRANCH_NAME in env')
}

const BRANCH_NAME = process.env.APP_BRANCH_NAME

export function getBranchName(): string {
  return BRANCH_NAME
}
