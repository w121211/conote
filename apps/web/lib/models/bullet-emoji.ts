import { BulletEmoji, BulletEmojiCount, BulletEmojiLike, LikeChoice } from '@prisma/client'
import { EmojiCode } from 'graphql-let/__generated__/__types__'
import prisma from '../prisma'
import { EmojiLike } from './emoji-like'

export const BulletEmojiModel = {
  async create({ bulletId, code, userId }: { bulletId: string; code: EmojiCode; userId: string }): Promise<{
    emoji: BulletEmoji & { count: BulletEmojiCount }
    like: BulletEmojiLike
  }> {
    const found = await prisma.bulletEmoji.findUnique({
      where: { bulletId_code: { bulletId, code } },
    })
    if (found) {
      // emoji existed, upsert like
      const { like, count } = await this.upsertLike({ choice: LikeChoice.UP, bulletEmojiId: found.id, userId })
      return {
        emoji: { ...found, count },
        like,
      }
    }
    // emoji not exist, create emoji & like
    const emoji = await prisma.bulletEmoji.create({
      data: {
        bullet: { connect: { id: bulletId } },
        code,
        count: { create: {} },
      },
    })
    const { like, count } = await this.upsertLike({ choice: LikeChoice.UP, bulletEmojiId: emoji.id, userId })
    return {
      emoji: { ...emoji, count },
      like,
    }
  },

  async upsertLike({
    choice,
    bulletEmojiId,
    userId,
  }: {
    choice: LikeChoice
    bulletEmojiId: string
    userId: string
  }): Promise<{
    like: BulletEmojiLike
    count: BulletEmojiCount
  }> {
    const curCount = await prisma.bulletEmojiCount.findUnique({
      where: { bulletEmojiId },
    })
    if (curCount === null) {
      throw 'Database unexpected error'
    }

    const curLike = await prisma.bulletEmojiLike.findUnique({
      where: {
        userId_bulletEmojiId: { userId, bulletEmojiId },
      },
    })
    const nextLike = curLike
      ? await prisma.bulletEmojiLike.update({
          data: { choice },
          where: { id: curLike.id },
        })
      : await prisma.bulletEmojiLike.create({
          data: {
            choice,
            bulletEmoji: { connect: { id: bulletEmojiId } },
            user: { connect: { id: userId } },
          },
        })

    const { dUp, dDown } = EmojiLike.compare(nextLike, curLike ?? undefined)
    const nextCount = await prisma.bulletEmojiCount.update({
      data: {
        nUps: curCount.nUps + dUp,
        nDowns: curCount.nDowns + dDown,
      },
      where: { id: curCount.id },
    })
    return { like: nextLike, count: nextCount }
  },
}
