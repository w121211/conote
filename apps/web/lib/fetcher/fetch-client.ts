import { MemoryCache } from './memory-cache'
import { tryFetch } from './vendors/index'

export type SrcType = 'VIDEO' | 'POST' | 'AUTHOR' | 'OTHER'

export type FetchResult = {
  domain: string
  finalUrl: string
  srcId?: string
  srcType: SrcType

  // metascraper
  authorId?: string
  authorName?: string
  date?: string
  description?: string
  lang?: string
  title?: string

  // ./packages/scraper
  keywords?: string[]
  tickers?: string[]
  error?: string
}

export class FetchClient {
  private cache: MemoryCache | null = null // cache support

  constructor(localCachePath: string | null = null) {
    if (localCachePath) {
      this.cache = new MemoryCache(localCachePath)
    }
  }

  public async fetch(url: string): Promise<FetchResult & { fromCache?: true }> {
    if (this.cache) {
      try {
        return {
          ...this.cache.get(url),
          fromCache: true,
        }
      } catch {
        const res = await tryFetch(url)
        this.cache.set(url, res)
        return res
      }
    }
    return await tryFetch(url)
  }

  public dump(): void {
    if (this.cache) {
      this.cache.dump()
    }
  }
}
