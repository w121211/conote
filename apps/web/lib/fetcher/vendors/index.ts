import { base } from './base'
import { youtube } from './youtube'
import { FetchResult } from '../fetch-client'

export type DomainFetchFunction = (url: string, domain?: string) => Promise<FetchResult>

export class DomainNotFitError extends Error {
  constructor(message?: string) {
    super(message) // 'Error' breaks prototype chain here
    Object.setPrototypeOf(this, DomainNotFitError.prototype) // restore prototype chain
  }
}

// fallback 依序嘗試執行，若失敗則嘗試下一個
const fetchers: DomainFetchFunction[] = [youtube, base]

export async function tryFetch(url: string): Promise<FetchResult> {
  for (const fetcher of fetchers) {
    try {
      // eslint-disable-next-line no-await-in-loop
      return await fetcher(url)
    } catch (err) {
      if (err instanceof DomainNotFitError) {
        continue
      } else {
        console.error(err)
        break
      }
    }
  }
  throw new Error(`Fetch error: ${url}`)
}
