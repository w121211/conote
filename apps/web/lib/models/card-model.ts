import { Card, Link, Sym, SymType } from '@prisma/client'
import { FetchClient } from '../fetcher/fetch-client'
import prisma from '../prisma'
import { CardStateModel, CardStateParsed } from './card-state-model'
import { LinkService } from './link-model'
import { SymModel } from './sym-model'

export type CardPrarsed = Omit<Card, 'meta'> & {
  meta: CardMeta
  sym: Sym
  state: CardStateParsed | null
  link: Link | null
}

export type CardMeta = {
  template?: 'webpage' | 'ticker' | 'topic' | 'vs'
  redirects?: string[]
  duplicates?: string[]
  url?: string
  author?: string
  publishedAt?: string
  description?: string
  keywords?: string[]
  // lang?: string
  title?: string
  tickers?: string[]
}

export type RowCard = Card & { sym: Sym; link: Link | null }

export const CardModel = {
  async get(id: string): Promise<null | CardPrarsed> {
    const card = await prisma.card.findUnique({
      where: { id },
      include: {
        link: true,
        sym: true,
      },
    })
    if (card) {
      return {
        ...card,
        meta: card.meta as unknown as CardMeta,
        state: await CardStateModel.getLastCardState(card.id),
      }
    }
    return null
  },

  /**
   * Get card by symbol
   *
   * @throw symbol parse error
   */
  async getBySymbol(symbol: string): Promise<CardPrarsed | null> {
    const parsed = SymModel.parse(symbol)
    const sym = await prisma.sym.findUnique({
      where: { symbol: parsed.symbol },
      include: { card: { include: { link: true } } },
    })
    if (sym?.card) {
      const { card, ...restSym } = sym
      const state = await CardStateModel.getLastCardState(card.id)
      // if (state === null) {
      //   throw '[conote-web] symbol-card state cannot be null'
      // }
      return {
        ...card,
        meta: card.meta as unknown as CardMeta,
        state,
        sym: restSym,
      }
    }
    return null
  },

  /**
   * Get or create a webpage-card by URL, if the URL not found in database, will try to scrape and then store
   * Newly created webpage-card will not have card-state and remains null.
   */
  async getOrCreateByUrl({
    scraper,
    url,
  }: {
    scraper?: FetchClient
    url: string
  }): Promise<CardPrarsed & { link: Link }> {
    const [link, { fetchResult }] = await LinkService.getOrCreateLink({ scraper, url })
    if (link.card) {
      const { card, ...linkRest } = link
      return {
        ...card,
        link: linkRest,
        meta: card.meta as unknown as CardMeta,
        state: await CardStateModel.getLastCardState(card.id),
      }
    }

    // Card not found, create one
    const meta: CardMeta = {
      template: 'webpage',
      url: link.url,
      author: fetchResult?.authorName,
      publishedAt: fetchResult?.date,
      description: fetchResult?.description,
      keywords: fetchResult?.keywords,
      title: fetchResult?.title,
      tickers: fetchResult?.tickers,
    }
    const card = await prisma.card.create({
      data: {
        link: { connect: { id: link.id } },
        meta,
        sym: {
          create: {
            type: SymType.URL,
            symbol: LinkService.toSymbol(link),
          },
        },
      },
      include: { sym: true },
    })
    return {
      ...card,
      link,
      meta,
      state: null,
    }
  },

  parse(card: RowCard, cardState: CardStateParsed | null = null): CardPrarsed {
    const { link, sym } = card
    return {
      ...card,
      meta: card.meta as unknown as CardMeta,
      state: cardState,
      sym,
      link,
    }
  },
}
