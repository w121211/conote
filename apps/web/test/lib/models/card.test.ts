/**
 * yarn run test-lib card.test
 */
/* eslint-disable no-await-in-loop */
// import _ from 'lodash'
import { readFileSync } from 'fs'
import { resolve } from 'path'
import { Card, PrismaClient } from '@prisma/client'
// import { getOrCreateCardBySymbol, getOrCreateCardByLink, CardMeta } from '../card'
import { FetchClient } from '../../../lib/fetcher/fetcher'
import { CardModel } from '../../../lib/models/card-model'
import prisma from '../../../lib/prisma'
import { clean } from '../../test-helpers'

const sampleFilepath = resolve(__dirname, '__samples__', 'common.txt')

// const params: CardTemplateParams = {
//   name: '$AAA',
//   template: templateTicker.name,
//   title: 'AAA Company',
//   ticker: '$AAA',
// }

// const input0: BulletInput = {
//   head: '111',
//   body: '111-111',
//   op: 'create',
//   children: [
//     { head: '222', body: '222-222', op: 'create' },
//     {
//       head: '333',
//       op: 'create',
//       children: [
//         { head: '444', op: 'create' },
//         { head: '555', op: 'create' },
//       ],
//     },
//   ],
// }

let fetcher: FetchClient
let card: Card

// beforeAll(async () => {
//   console.log('Writing required data into database')
//   const _prisma = new PrismaClient()
//   await createTestUsers(_prisma)
//   card = await _prisma.card.create({
//     data: {
//       meta?: JsonNullValueInput | InputJsonValue
//     createdAt?: Date | string
//     updatedAt?: Date | string
//     sym: SymCreateNestedOneWithoutCardInput
//     link?: LinkCreateNestedOneWithoutCardInput
//     states?: CardStateCreateNestedManyWithoutCardInput
//     emojis?: CardEmojiCreateNestedManyWithoutCardInput
//       type: CardTy.TICKER,
//       name: params.name,
//       link: {
//         create: {
//           url: symbolToUrl(params.name),
//           domain: SYMBOL_DOMAIN,
//           srcType: SrcType.OTHER,
//         },
//       },
//     },
//   })
//   await _prisma.$disconnect()
//   console.log('Setting up a fetch-client')
//   fetcher = new FetchClient(resolve(__dirname, 'local-cache.dump.json'))
// })

afterAll(async () => {
  fetcher.dump()
  await prisma.$disconnect()
})

// test('replaceBulletInput()', () => {
//   expect(clean(replaceBulletInput(templateTicker.body, params))).toMatchSnapshot()
// })

// test('runBulletInputOp()', async () => {
//   const root0 = await runBulletInputOp({
//     cardId: card.id,
//     input: input0,
//     // prevDict,
//     timestamp: 111111111,
//     userId: TESTUSERS[0].id,
//   })
//   expect(clean(root0)).toMatchSnapshot()
// })

// test('createCardBody()', async () => {
//   const body0 = await createCardBody({ cardId: card.id, bulletInputRoot: input0 }, TESTUSERS[0].id)
//   expect(clean({ ...body0, content: omitDeep(JSON.parse(body0.content), ['timestamp']) })).toMatchSnapshot()

//   const input1: BulletInput = {
//     id: 6,
//     head: '111+++',
//     body: '111-111+++',
//     op: 'update',
//     children: [
//       { id: 7, head: '222', body: '222-222', op: 'delete' },
//       {
//         id: 8,
//         head: '333',
//         children: [{ id: 9, head: '444' }],
//       },
//       { id: 10, head: '555', op: 'move' },
//       { head: '666', op: 'create' },
//     ],
//   }

//   const body1 = await createCardBody({ cardId: card.id, bulletInputRoot: input1 }, TESTUSERS[1].id)
//   console.log(body1)
//   expect(clean({ ...body1, content: omitDeep(JSON.parse(body1.content), ['timestamp']) })).toMatchSnapshot()
// })

// test('createCard()', async () => {
//   const { card, body } = await createCard({ name: '$BBB', template: templateTicker, title: 'BBB Company' })
//   expect(clean(card)).toMatchSnapshot()
//   expect(clean({ ...body, content: omitDeep(JSON.parse(body.content), ['timestamp']) })).toMatchSnapshot()
// })

it.each(['https://zhuanlan.zhihu.com/p/75120221', 'https://www.mobile01.com/topicdetail.php?f=793&t=6520838'])(
  'getOrCreateByUrl()',
  async url => {
    const card = await CardModel.getOrCreateByUrl({ url: url })
    expect(clean(card)).toMatchSnapshot()
  },
)

// test('create a symbol card', async () => {
//   const card = await createCard('$AA')
//   // const discussBoard = await prisma.comment.findUnique({
//   //   where: { id: (card.meta as unknown as CardMeta).commentId },
//   // })
//   expect(clean(card)).toMatchSnapshot()
//   // expect(clean(discussBoard)).toMatchSnapshot()
//   // expect(_clean(editor.getMarkerlines())).toMatchSnapshot()
// })

// test('create a symbol card body', async () => {
//   const card = await getOrCreateCardBySymbol('$AA')
//   const editor = new Editor(card.body.text)
//   editor.setText(`[?] <BUY> @30
// [+]
// 111 111
// 222 222`)
//   editor.flush()

//   await createCardBody(card, editor, TESTUSERS[0].id)
//   const temp = await prisma.cocard.findUnique({
//     where: { id: card.id },
//     include: { body: true },
//   })
//   expect(clean(temp)).toMatchSnapshot()
//   // expect(_clean(editor.getMarkerlines())).toMatchSnapshot()
// })

// test('create a blank web card', async () => {
//   const [link] = await getOrCreateLink('http://test1.com')
//   const card = await getOrCreateCardByLink(link)
//   expect(clean(card)).toMatchSnapshot()
// })

// test('edit a web card', async () => {
//   const card = await prisma.cocard.findUnique({
//     where: { linkUrl: 'http://test1.com' },
//     include: { body: true },
//   })
//   if (card === null) throw new Error()

//   const editor = new Editor(card.body.text)
//   editor.setText(`[*]
// 2021年會有進一步財政刺激
// 歷年有1月效應，1月份呈現反彈，但不認為`)
//   editor.flush()

//   await createCardBody(card, editor, TESTUSERS[0].id)
//   const temp = await prisma.cocard.findUnique({
//     where: { id: card.id },
//     include: { body: true },
//   })
//   expect(clean(temp)).toMatchSnapshot()
// })

// test('create a nested web-card', async () => {
//   const [link, { fetchResult }] = await getOrCreateLink('http://test2.com')
//   const card = await getOrCreateCardByLink(link)

//   const editor = new Editor(
//     card.body.text,
//     card.body.meta as unknown as Markerline[],
//     'http://test2.com',
//     TESTOAUTHORS[0].name,
//   )
//   editor.setText(`$CC
// [?] <BUY> @30
// [+]
// 111 111
// 222 222

// $DD
// [?] <SELL> @30`)
//   editor.flush()

//   for (const [cardlabel, markerlines] of editor.getNestedMarkerlines()) {
//     const nestedCard = await getOrCreateCardBySymbol(cardlabel.symbol)
//     const nestedEditor = new Editor(
//       nestedCard.body.text,
//       nestedCard.body.meta as unknown as Markerline[],
//       link.url,
//       TESTOAUTHORS[0].name,
//     )
//     nestedEditor.setMarkerlinesToInsert(markerlines.filter(e => e.new && !e.neatReply))
//     nestedEditor.flush()
//     console.log(nestedEditor.getText())
//     console.log(nestedEditor.getMarkerlines())

//     await createCardBody(nestedCard, nestedEditor, TESTUSERS[1].id)
//   }

//   await createCardBody(card, editor, TESTUSERS[1].id)

//   expect((await getOrCreateCardByLink(link)).body.text).toMatchSnapshot()
//   expect(clean((await getOrCreateCardByLink(link)).body.meta as unknown as Record<string, unknown>)).toMatchSnapshot()
//   expect((await getOrCreateCardBySymbol('$CC')).body.text).toMatchSnapshot()
//   expect(
//     clean((await getOrCreateCardBySymbol('$CC')).body.meta as unknown as Record<string, unknown>),
//   ).toMatchSnapshot()
//   expect((await getOrCreateCardBySymbol('$DD')).body.text).toMatchSnapshot()
//   expect(
//     clean((await getOrCreateCardBySymbol('$DD')).body.meta as unknown as Record<string, unknown>),
//   ).toMatchSnapshot()

//   // await createCardBody(card, editor, TESTUSERS[1].id)

//   // const rootCard = await prisma.cocard.findUnique({
//   //   where: { id: card.id },
//   //   include: { body: true },
//   // })
//   // expect(omitDeep(rootCard ?? {}, ['createdAt', 'updatedAt'])).toMatchSnapshot()

//   // const nestedCard = await prisma.cocard.findUnique({
//   //   where: { linkUrl: '//$ABBV' },
//   //   include: { body: true },
//   // })
//   // if (nestedCard === null) throw new Error()
//   // expect(omitDeep(nestedCard ?? {}, ['createdAt', 'updatedAt'])).toMatchSnapshot()

//   // const markerlines = new TextEditor(nestedCard.body.text).getMarkerlines()
//   // expect(markerlines).toMatchSnapshot()
// })

// test.each(
//   splitByUrl(readFileSync(sampleFilepath, { encoding: 'utf8' }))
//     .filter((e): e is [string, string] => e[0] !== undefined)
//     .map(e => [e[0].trim(), e[1].trim()]),
// )('Create web card from common.txt', async (url: string, body: string) => {
//   if (url === undefined) return

//   const [link] = await getOrCreateLink(url, fetcher)
//   const card = await getOrCreateCardByLink(link)
//   expect((await createWebCardBody(card.id, body, TESTUSERS[0].id)).text).toMatchSnapshot()
//   // expect((await createWebCardBody(card.id, body, TESTUSERS[0].id)).meta).toMatchSnapshot()
// })
