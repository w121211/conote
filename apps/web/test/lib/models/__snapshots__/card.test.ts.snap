// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getOrCreateByUrl() 1`] = `
Object {
  "link": Object {
    "author": Object {
      "meta": Object {},
      "name": "梁上尘梁上尘土",
    },
    "authorId": "cky5ia5ta00048gsbwql5j5qc",
    "domain": "zhuanlan.zhihu.com",
    "scraped": Object {
      "authorName": "梁上尘梁上尘土",
      "date": "2019-07-25T11:35:00.000Z",
      "description": "现在比较热门的话题碳化硅，和小编一起看看碳化硅： 半导体材料经过几十年的发展，第一代硅材料半导体已经接近完美晶体，对于硅材料的研究也非常透彻。基于硅材料上器件的设计和开发也经过了许多代的结构和工艺优…",
      "domain": "zhuanlan.zhihu.com",
      "finalUrl": "https://zhuanlan.zhihu.com/p/75120221",
      "lang": "zh",
      "srcType": "OTHER",
      "title": "三分钟了解第三代半导体材料：碳化硅（SiC）",
    },
    "url": "https://zhuanlan.zhihu.com/p/75120221",
  },
  "linkId": "cky5ia5u500138gsbw46qx8uw",
  "meta": Object {
    "author": "梁上尘梁上尘土",
    "description": "现在比较热门的话题碳化硅，和小编一起看看碳化硅： 半导体材料经过几十年的发展，第一代硅材料半导体已经接近完美晶体，对于硅材料的研究也非常透彻。基于硅材料上器件的设计和开发也经过了许多代的结构和工艺优…",
    "publishedAt": "2019-07-25T11:35:00.000Z",
    "template": "webpage",
    "title": "三分钟了解第三代半导体材料：碳化硅（SiC）",
    "url": "https://zhuanlan.zhihu.com/p/75120221",
  },
  "state": null,
  "sym": Object {
    "symbol": "@https://zhuanlan.zhihu.com/p/75120221",
    "type": "URL",
  },
}
`;

exports[`getOrCreateByUrl() 2`] = `
Object {
  "link": Object {
    "author": Object {
      "meta": Object {},
      "name": "Evelyn_Tsai",
    },
    "authorId": "cky5owuyp00168bsb49nppwu5",
    "domain": "www.mobile01.com",
    "scraped": Object {
      "authorName": "Evelyn_Tsai",
      "date": "2022-01-06T11:12:46.000Z",
      "description": "興櫃股國鼎生技（4132）這一個星期三（2022年1月5日）召開重訊宣布其所自行研發（研發中）的新型冠狀病毒肺炎住院口服用藥－Antroquinonol二期臨床試驗解盲達標，預計於今年度四月份的時候向美國食品藥物管理局（FDA）遞交二期臨床試驗的分析數據以及研發相關文件並申請緊(股票 / ETF 第1頁)",
      "domain": "www.mobile01.com",
      "finalUrl": "https://www.mobile01.com/topicdetail.php?f=793&t=6520838",
      "keywords": Array [
        "股票 / ETF",
        "理財",
      ],
      "lang": "zh",
      "srcType": "OTHER",
      "title": "國鼎新冠口服藥二期解盲成功／股價狂噴67%，開盤８分鐘就嗨到觸熔斷 - Mobile01",
    },
    "url": "https://www.mobile01.com/topicdetail.php?f=793&t=6520838",
  },
  "linkId": "cky5owuzy00258bsb4gpc25fo",
  "meta": Object {
    "author": "Evelyn_Tsai",
    "description": "興櫃股國鼎生技（4132）這一個星期三（2022年1月5日）召開重訊宣布其所自行研發（研發中）的新型冠狀病毒肺炎住院口服用藥－Antroquinonol二期臨床試驗解盲達標，預計於今年度四月份的時候向美國食品藥物管理局（FDA）遞交二期臨床試驗的分析數據以及研發相關文件並申請緊(股票 / ETF 第1頁)",
    "keywords": Array [
      "股票 / ETF",
      "理財",
    ],
    "publishedAt": "2022-01-06T11:12:46.000Z",
    "template": "webpage",
    "title": "國鼎新冠口服藥二期解盲成功／股價狂噴67%，開盤８分鐘就嗨到觸熔斷 - Mobile01",
    "url": "https://www.mobile01.com/topicdetail.php?f=793&t=6520838",
  },
  "state": null,
  "sym": Object {
    "symbol": "@https://www.mobile01.com/topicdetail.php?f=793&t=6520838",
    "type": "URL",
  },
}
`;
